File: src/lib/research/jurisdictionResolve.ts
// Orchestrates the full local-jurisdiction resolution for a facility location.
//
// Given a county (and optional city), it composes the three resolvers into one
// answer the run pipeline can act on:
//   - resolveAirDistrict   -> which of the 35 air districts cover the county
//   - resolveJurisdictionSkills -> the researched county/city skill folders
//   - (resolveWaterBoard folds in here once the county->region map lands)
//
// Every uncertainty is surfaced as an explicit gap, never papered over with a
// guessed authority. A county that spans multiple air districts can't be pinned
// on the county name alone (needs sub-county geometry) -> that's a gap. A county
// or city with no researched skill folder -> that's a gap. An unknown county ->
// that's a gap and an empty stack. The run treats gaps as fail-closed missing
// facts, so the report says "needs review" instead of citing the wrong agency.

import { resolveAirDistrict, resolveWaterBoard, type AirDistrict, type RegionalWaterBoard } from "./jurisdictionRegistry";
import { resolveJurisdictionSkills, type JurisdictionSkill } from "./jurisdictionSkills";
import type { ScopePack } from "./types";

export type ResolvedJurisdiction = {
  county: string | null;
  city: string | null;
  airDistricts: AirDistrict[];
  airNeedsGeometry: boolean;
  waterBoards: RegionalWaterBoard[];
  waterNeedsGeometry: boolean;
  countySkill: JurisdictionSkill | null;
  citySkill: JurisdictionSkill | null;
  // Human-readable resolved authority names for ScopePack.facility.jurisdiction_stack.
  stack: string[];
  // Honest gaps: unknown county, county spans multiple air districts (needs
  // geometry), or an unresearched county/city skill folder.
  gaps: string[];
};

export function resolveJurisdiction(facility: { county: string | null; city: string | null }): ResolvedJurisdiction {
  const county = facility.county;
  const city = facility.city;

  if (!county) {
    return {
      county: null,
      city,
      airDistricts: [],
      airNeedsGeometry: false,
      waterBoards: [],
      waterNeedsGeometry: false,
      countySkill: null,
      citySkill: null,
      stack: [],
      gaps: ["location:county_unknown"],
    };
  }

  const air = resolveAirDistrict(county);
  const water = resolveWaterBoard(county);
  const skills = resolveJurisdictionSkills({ county, ...(city ? { city } : {}) });

  const gaps: string[] = [...skills.gaps];
  // More than one district/region covers the county -> can't resolve on the
  // county name alone. Don't pick one arbitrarily; flag that sub-county geometry
  // is needed. Zero matches -> flag the missing authority.
  if (air.needsGeometry) gaps.push(`air_geometry:${county}`);
  if (air.districts.length === 0) gaps.push(`air_district:${county}`);
  if (water.needsGeometry) gaps.push(`water_geometry:${county}`);
  if (water.boards.length === 0) gaps.push(`water_board:${county}`);

  const stack: string[] = [];
  for (const d of air.districts) stack.push(d.name);
  for (const b of water.boards) stack.push(b.name);
  if (skills.county) stack.push(`${county} County local programs (CUPA / fire / building)`);
  if (skills.city && city) stack.push(`${city} local programs`);

  return {
    county,
    city,
    airDistricts: air.districts,
    airNeedsGeometry: air.needsGeometry,
    waterBoards: water.boards,
    waterNeedsGeometry: water.needsGeometry,
    countySkill: skills.county,
    citySkill: skills.city,
    stack,
    gaps,
  };
}

// Which coverage dimensions a jurisdiction gap blocks. A gap means we don't know
// the controlling authority, so any determination depending on it is unsafe.
function blocksForGap(gap: string): string[] {
  if (gap.startsWith("air_")) return ["air"];
  if (gap.startsWith("water_board")) return ["stormwater", "wastewater"];
  if (gap.startsWith("location:")) return ["air", "stormwater", "hazmat", "waste", "wastewater"];
  // county:/city: skill folders carry the local CUPA / fire / building authority.
  return ["hazmat", "waste"];
}

// Fold jurisdiction resolution into a ScopePack (pure — returns a new pack):
//  - when a real resolution exists, replace the conservative default
//    jurisdiction_stack with the actual resolved authorities;
//  - turn every resolution gap into a fail-closed missing_fact so the run
//    surfaces "needs review" rather than citing a guessed agency.
export function applyJurisdictionToScope(scope: ScopePack): ScopePack {
  const resolved = resolveJurisdiction(scope.facility);

  const jurisdiction_stack =
    resolved.stack.length > 0 ? resolved.stack : scope.facility.jurisdiction_stack;

  const missing_facts = scope.missing_facts.map((m) => ({ ...m, blocks: [...m.blocks] }));
  for (const gap of resolved.gaps) {
    const field = `jurisdiction.${gap}`;
    if (missing_facts.some((m) => m.field === field)) continue;
    missing_facts.push({
      field,
      why_needed:
        "Local jurisdiction could not be resolved to a single controlling authority; " +
        "the responsible agency and its adopted rules must be confirmed before any determination.",
      blocks: blocksForGap(gap),
    });
  }

  return {
    ...scope,
    facility: { ...scope.facility, jurisdiction_stack },
    missing_facts,
  };
}

File: src/research_core/jurisdiction_resolve.py
from __future__ import annotations

from dataclasses import dataclass

from research_core.jurisdiction_registry import (
    AirDistrict,
    RegionalWaterBoard,
    resolve_air_district,
    resolve_water_board,
)
from research_core.jurisdiction_skills import (
    JurisdictionSkill,
    resolve_jurisdiction_skills,
)
from research_core.models import MissingFact, ScopePack


@dataclass(frozen=True)
class ResolvedJurisdiction:
    county: str | None
    city: str | None
    air_districts: tuple[AirDistrict, ...]
    air_needs_geometry: bool
    water_boards: tuple[RegionalWaterBoard, ...]
    water_needs_geometry: bool
    county_skill: JurisdictionSkill | None
    city_skill: JurisdictionSkill | None
    stack: tuple[str, ...]
    gaps: tuple[str, ...]


ALL_JURISDICTION_BLOCKS = (
    "air",
    "stormwater",
    "hazmat",
    "waste",
    "wastewater",
    "land_use",
    "fire_code",
)


def resolve_jurisdiction(
    facility: dict[str, str | None],
) -> ResolvedJurisdiction:
    county = facility.get("county")
    city = facility.get("city")

    if not county:
        return ResolvedJurisdiction(
            county=None,
            city=city,
            air_districts=(),
            air_needs_geometry=False,
            water_boards=(),
            water_needs_geometry=False,
            county_skill=None,
            city_skill=None,
            stack=(),
            gaps=("location:county_unknown",),
        )

    air = resolve_air_district(county)
    water = resolve_water_board(county)
    skills = resolve_jurisdiction_skills(
        {"county": county, **({"city": city} if city else {})}
    )

    gaps = list(skills.gaps)
    if air.needs_geometry:
        gaps.append(f"air_geometry:{county}")
    if len(air.districts) == 0:
        gaps.append(f"air_district:{county}")
    if water.needs_geometry:
        gaps.append(f"water_geometry:{county}")
    if len(water.boards) == 0:
        gaps.append(f"water_board:{county}")

    stack: list[str] = []
    stack.extend(district.name for district in air.districts)
    stack.extend(board.name for board in water.boards)
    if skills.county:
        stack.append(f"{county} County local programs (CUPA / fire / building)")
    if skills.city and city:
        stack.append(f"{city} local programs")

    return ResolvedJurisdiction(
        county=county,
        city=city,
        air_districts=air.districts,
        air_needs_geometry=air.needs_geometry,
        water_boards=water.boards,
        water_needs_geometry=water.needs_geometry,
        county_skill=skills.county,
        city_skill=skills.city,
        stack=tuple(stack),
        gaps=tuple(gaps),
    )


def jurisdiction_context_for(facility: dict[str, str | None]) -> str:
    resolved = resolve_jurisdiction(facility)
    parts: list[str] = []

    if resolved.stack:
        parts.append("Resolved controlling authorities for this location:")
        for name in resolved.stack:
            parts.append(f"  - {name}")

    if resolved.county_skill:
        parts.append(
            "\n## County local-authority reference\n"
            + resolved.county_skill.content.strip()
        )
    if resolved.city_skill:
        parts.append(
            "\n## City local-authority reference\n"
            + resolved.city_skill.content.strip()
        )

    if resolved.gaps:
        parts.append(
            "\nUNRESOLVED jurisdiction levels - do NOT assume an authority for these; "
            "treat any dependent determination as needs_review until confirmed: "
            + ", ".join(resolved.gaps)
        )

    if parts:
        return "\n".join(parts)
    return (
        "Jurisdiction unresolved - no county provided; confirm the controlling "
        "authority before any local determination."
    )


def _blocks_for_gap(gap: str) -> list[str]:
    if gap.startswith("air_"):
        return ["air"]
    if gap.startswith("water_"):
        return ["stormwater", "wastewater"]
    if gap.startswith("location:"):
        return list(ALL_JURISDICTION_BLOCKS)
    if gap.startswith("county:") or gap.startswith("city:"):
        return ["hazmat", "waste", "land_use", "fire_code"]
    return list(ALL_JURISDICTION_BLOCKS)


def _model_copy(model, **updates):
    if hasattr(model, "model_copy"):
        return model.model_copy(update=updates, deep=True)
    return model.copy(update=updates, deep=True)


def apply_jurisdiction_to_scope(scope: ScopePack) -> ScopePack:
    resolved = resolve_jurisdiction(
        {"county": scope.facility.county, "city": scope.facility.city}
    )
    jurisdiction_stack = (
        resolved.stack if resolved.stack else scope.facility.jurisdiction_stack
    )

    missing_facts = [
        _model_copy(missing_fact, blocks=list(missing_fact.blocks))
        for missing_fact in scope.missing_facts
    ]
    existing_fields = {missing_fact.field for missing_fact in missing_facts}
    for gap in resolved.gaps:
        field = f"jurisdiction.{gap}"
        if field in existing_fields:
            continue
        missing_facts.append(
            MissingFact(
                field=field,
                why_needed=(
                    "Local jurisdiction could not be resolved to a single "
                    "controlling authority; the responsible agency and its "
                    "adopted rules must be confirmed before any determination."
                ),
                blocks=_blocks_for_gap(gap),
            )
        )
        existing_fields.add(field)

    facility = _model_copy(
        scope.facility,
        jurisdiction_stack=list(jurisdiction_stack),
    )
    return _model_copy(scope, facility=facility, missing_facts=missing_facts)
